From 031207e08f93bc8d8a81a9d411852fd25c2682ef Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:31:59 -0400
Subject: [scsi] hpsa: define extended_report_lun_entry data structure

Message-id: <1410186734-19986-69-git-send-email-thenzl@redhat.com>
Patchwork-id: 92104
O-Subject: [RHEL6.6 e-stor PATCH 68/83] hpsa: define extended_report_lun_entry data structure
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Reviewed-by: Joe Handzik <joseph.t.handzik@hp.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit d5b5d964561c096bc262cce12a06aff76f17d7d9)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 7a20000..a81e891 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -2875,26 +2875,20 @@ static int hpsa_get_pdisk_of_ioaccel2(struct ctlr_info *h,
  nphysicals = be32_to_cpu(*((__be32 *)physicals->LUNListLength)) /
        responsesize;
 
-
  /* find ioaccel2 handle in list of physicals: */
  for (i = 0; i < nphysicals; i++) {
+  struct ext_report_lun_entry *entry = &physicals->LUN[i];
+
   /* handle is in bytes 28-31 of each lun */
-  if (memcmp(&((struct ReportExtendedLUNdata *)
-    physicals)->LUN[i][20], &find, 4) != 0) {
+  if (entry->ioaccel_handle != find)
    continue; /* didn't match */
-  }
   found = 1;
-  memcpy(scsi3addr, &((struct ReportExtendedLUNdata *)
-     physicals)->LUN[i][0], 8);
+  memcpy(scsi3addr, entry->lunid, 8);
   if (h->raid_offload_debug > 0)
    dev_info(&h->pdev->dev,
-    "%s: Searched h=0x%08x, Found h=0x%08x, scsiaddr 0x%02x%02x%02x%02x%02x%02x%02x%02x\n",
+    "%s: Searched h=0x%08x, Found h=0x%08x, scsiaddr 0x%8phN\n",
     __func__, find,
-    ((struct ReportExtendedLUNdata *)
-     physicals)->LUN[i][20],
-    scsi3addr[0], scsi3addr[1], scsi3addr[2],
-    scsi3addr[3], scsi3addr[4], scsi3addr[5],
-    scsi3addr[6], scsi3addr[7]);
+    entry->ioaccel_handle, scsi3addr);
   break; /* found it */
  }
 
@@ -2979,7 +2973,8 @@ u8 *figure_lunaddrbytes(struct ctlr_info *h, int raid_ctlr_position, int i,
   return RAID_CTLR_LUNID;
 
  if (i < logicals_start)
-  return &physdev_list->LUN[i - (raid_ctlr_position == 0)][0];
+  return &physdev_list->LUN[i -
+    (raid_ctlr_position == 0)].lunid[0];
 
  if (i < last_device)
   return &logdev_list->LUN[i - nphysicals -
diff --git a/drivers/scsi/hpsa_cmd.h b/drivers/scsi/hpsa_cmd.h
index 104b67b..649b463 100644
--- a/drivers/scsi/hpsa_cmd.h
+++ b/drivers/scsi/hpsa_cmd.h
@@ -238,11 +238,21 @@ struct ReportLUNdata {
  u8 LUN[HPSA_MAX_LUN][8];
 };
 
+struct ext_report_lun_entry {
+ u8 lunid[8];
+ u8 wwid[8];
+ u8 device_type;
+ u8 device_flags;
+ u8 lun_count; /* multi-lun device, how many luns */
+ u8 redundant_paths;
+ u32 ioaccel_handle; /* ioaccel1 only uses lower 16 bits */
+};
+
 struct ReportExtendedLUNdata {
  u8 LUNListLength[4];
  u8 extended_response_flag;
  u8 reserved[3];
- u8 LUN[HPSA_MAX_LUN][24];
+ struct ext_report_lun_entry LUN[HPSA_MAX_LUN];
 };
 
 struct SenseSubsystem_info {
-- 
1.7.1